File: sdk/go/plugin-sdk-go/server_test.go
package architectplugin

import (
	"bytes"
	"context"
	"encoding/json"
	"testing"
)

type stubPlugin struct{}

func (stubPlugin) Init(_ context.Context, config map[string]any) (InitResult, error) {
	if config["plugin"] != true {
		return InitResult{}, nil
	}
	return InitResult{OK: true, Name: "example", Version: "1.0.0"}, nil
}

func (stubPlugin) ListTasks(_ context.Context) ([]TaskDescriptor, error) {
	return []TaskDescriptor{{ID: "build", Description: "Build"}}, nil
}

func (stubPlugin) ExecuteTask(_ context.Context, _ ExecuteTaskParams, writer TaskEventWriter) (int, error) {
	if err := writer.Output("starting"); err != nil {
		return 1, err
	}
	if err := writer.Progress(0.5); err != nil {
		return 1, err
	}
	return 0, nil
}

func TestServerHandleInitAndListTasks(t *testing.T) {
	var output bytes.Buffer
	server := NewServer(stubPlugin{}).WithIO(nil, &output, func(int) {})

	if err := server.HandleRequest(context.Background(), Request{
		JSONRPC: JSONRPCVersion,
		ID:      1,
		Method:  MethodInit,
		Params: InitParams{
			Config:          map[string]any{"plugin": true},
			ProtocolVersion: APPVersion,
		},
	}); err != nil {
		t.Fatalf("handle init: %v", err)
	}

	if err := server.HandleRequest(context.Background(), Request{
		JSONRPC: JSONRPCVersion,
		ID:      2,
		Method:  MethodListTasks,
	}); err != nil {
		t.Fatalf("handle listTasks: %v", err)
	}

	lines := bytes.Split(bytes.TrimSpace(output.Bytes()), []byte("\n"))
	if len(lines) != 2 {
		t.Fatalf("expected 2 responses, got %d", len(lines))
	}

	var initResponse Response
	if err := json.Unmarshal(lines[0], &initResponse); err != nil {
		t.Fatalf("decode init response: %v", err)
	}
	resultMap, ok := initResponse.Result.(map[string]any)
	if !ok {
		t.Fatalf("expected init result map, got %T", initResponse.Result)
	}
	if resultMap["name"] != "example" {
		t.Fatalf("expected plugin name example, got %v", resultMap["name"])
	}

	var taskResponse Response
	if err := json.Unmarshal(lines[1], &taskResponse); err != nil {
		t.Fatalf("decode listTasks response: %v", err)
	}
	tasks, ok := taskResponse.Result.([]any)
	if !ok || len(tasks) != 1 {
		t.Fatalf("expected one task, got %#v", taskResponse.Result)
	}
}

func TestServerHandleExecuteTask(t *testing.T) {
	var output bytes.Buffer
	server := NewServer(stubPlugin{}).WithIO(nil, &output, func(int) {})

	if err := server.HandleRequest(context.Background(), Request{
		JSONRPC: JSONRPCVersion,
		ID:      3,
		Method:  MethodExecuteTask,
		Params: ExecuteTaskParams{
			ID:   "build",
			Args: []string{"--scan"},
			Env:  map[string]string{"ARCHITECT_PROFILE": "default"},
		},
	}); err != nil {
		t.Fatalf("handle executeTask: %v", err)
	}

	lines := bytes.Split(bytes.TrimSpace(output.Bytes()), []byte("\n"))
	if len(lines) != 4 {
		t.Fatalf("expected 4 lines, got %d", len(lines))
	}

	var ack Response
	if err := json.Unmarshal(lines[0], &ack); err != nil {
		t.Fatalf("decode ack: %v", err)
	}
	if ack.ID == nil || *ack.ID != 3 {
		t.Fatalf("expected ack id 3, got %#v", ack.ID)
	}

	var outputEvent TaskEvent
	if err := json.Unmarshal(lines[1], &outputEvent); err != nil {
		t.Fatalf("decode output event: %v", err)
	}
	if outputEvent.Type != EventOutput || outputEvent.Data != "starting" {
		t.Fatalf("unexpected output event: %#v", outputEvent)
	}

	var progressEvent TaskEvent
	if err := json.Unmarshal(lines[2], &progressEvent); err != nil {
		t.Fatalf("decode progress event: %v", err)
	}
	if progressEvent.Type != EventProgress || progressEvent.Progress == nil || *progressEvent.Progress != 0.5 {
		t.Fatalf("unexpected progress event: %#v", progressEvent)
	}

	var completedEvent TaskEvent
	if err := json.Unmarshal(lines[3], &completedEvent); err != nil {
		t.Fatalf("decode completed event: %v", err)
	}
	if completedEvent.Type != EventCompleted || completedEvent.ExitCode == nil || *completedEvent.ExitCode != 0 {
		t.Fatalf("unexpected completed event: %#v", completedEvent)
	}
}

File: sdk/typescript/plugin-sdk/test/server.test.ts
import test from "node:test";
import assert from "node:assert/strict";
import { PassThrough, Writable } from "node:stream";
import { ArchitectProcessPlugin, PluginServer } from "../src";
import { APP_VERSION, JSON_RPC_VERSION, METHODS } from "../src/protocol";

class CaptureWritable extends Writable {
  private readonly chunks: string[] = [];

  _write(chunk: Buffer | string, _encoding: BufferEncoding, callback: (error?: Error | null) => void): void {
    this.chunks.push(chunk.toString());
    callback();
  }

  lines(): string[] {
    return this.chunks.join("").trim().split("\n").filter(Boolean);
  }
}

test("PluginServer handles init and listTasks", async () => {
  const output = new CaptureWritable();
  const plugin: ArchitectProcessPlugin = {
    async init(config) {
      assert.equal(config.plugin, true);
      return { ok: true, name: "example", version: "1.0.0" };
    },
    async listTasks() {
      return [{ id: "build", description: "Build" }];
    },
    async executeTask() {
      return 0;
    },
  };

  const server = new PluginServer(plugin, {
    input: new PassThrough(),
    output,
    exit: () => undefined,
  });

  await server.handleRequest({
    jsonrpc: JSON_RPC_VERSION,
    id: 1,
    method: METHODS.init,
    params: { config: { plugin: true }, protocol_version: APP_VERSION },
  });
  await server.handleRequest({
    jsonrpc: JSON_RPC_VERSION,
    id: 2,
    method: METHODS.listTasks,
  });

  const lines = output.lines().map((line) => JSON.parse(line));
  assert.equal(lines[0].result.ok, true);
  assert.equal(lines[0].result.name, "example");
  assert.deepEqual(lines[1].result, [{ id: "build", description: "Build" }]);
});

test("PluginServer acknowledges executeTask and streams events", async () => {
  const output = new CaptureWritable();
  const plugin: ArchitectProcessPlugin = {
    listTasks() {
      return [];
    },
    async executeTask(_request, writer) {
      await writer.output("starting");
      await writer.progress(0.5);
      return 0;
    },
  };

  const server = new PluginServer(plugin, {
    input: new PassThrough(),
    output,
    exit: () => undefined,
  });

  await server.handleRequest({
    jsonrpc: JSON_RPC_VERSION,
    id: 3,
    method: METHODS.executeTask,
    params: {
      id: "build",
      args: ["--scan"],
      env: { ARCHITECT_PROFILE: "default" },
    },
  });

  const lines = output.lines().map((line) => JSON.parse(line));
  assert.equal(lines[0].id, 3);
  assert.equal(lines[0].result, null);
  assert.equal(lines[1].type, "output");
  assert.equal(lines[1].data, "starting");
  assert.equal(lines[2].type, "progress");
  assert.equal(lines[2].progress, 0.5);
  assert.equal(lines[3].type, "completed");
  assert.equal(lines[3].exitCode, 0);
});